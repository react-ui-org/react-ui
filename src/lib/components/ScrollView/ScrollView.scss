// 1. Scrolling shadows are implemented as pseudo elements. This way we can customise them only
//    with custom properties.
//
// 2. Stack scrolling shadows  over viewport content while keeping the content interactive.
//
//    - `.scrollingShadows` is positioned absolutely over the `.root`, with auto `z-index` (this is
//      important!), and with `overflow: hidden` to clip the shadows (ie. its pseudo elements).
//    - The `.viewport` is in `.root`'s stacking context and remains interactive because its
//      `z-index` is higher than the auto `z-index` of `.scrollingShadows`.
//
// 3. Optional arrows are positioned relative to the `.root` and stacked on top of scrolling
//    shadows. They can be shifted outside the `ScrollView` area only because `overflow: hidden` is
//    **not** present at `.root`.
//
// 4. Make the `.content`'s bounding rectangle spread beyond the part visible through `.viewport`.
//
// 5. Prevent undesired vertical scrolling that may occur with tables inside.
//
// 6. Make `ScrollView` adjust to flexible layouts.
//
// 7. Hide content overflowing in the other direction because scrollbars would be unreachable under
//    scrolling shadows.

@use "../../styles/tools/accessibility";
@use "../../styles/tools/caret";
@use "../../styles/tools/reset";
@use "../../styles/tools/scrollbar";
@use "../../styles/tools/spacing";
@use "../../styles/tools/transition";
@use "theme";

$_arrow-inner-spacing: spacing.of(2);
$_arrow-outer-spacing: spacing.of(4);

.root {
    position: relative; // 2.
    display: flex;
    flex-direction: column;
    width: 100%;
}

// 1.
.scrollingShadows {
    position: absolute; // 2.
    width: 100%; // 2.
    height: 100%; // 2.
    overflow: hidden; // 2.

    &::before,
    &::after {
        @include transition.add((visibility, opacity, transform));

        content: "";
        position: absolute;
        z-index: 2; // 2.
        display: block;
        width: var(--rui-local-shadow-width);
        height: var(--rui-local-shadow-height);
        visibility: hidden;
        opacity: 0;
    }

    &::before {
        background: var(--rui-local-start-shadow-background);
        box-shadow: var(--rui-local-start-shadow-box-shadow);
    }

    &::after {
        background: var(--rui-local-end-shadow-background);
        box-shadow: var(--rui-local-end-shadow-box-shadow);
    }
}

.viewport {
    z-index: 1; // 2.
    width: 100%;
    scroll-behavior: smooth;
    -webkit-overflow-scrolling: touch;
}

.arrowPrev,
.arrowNext {
    @include reset.button();
    @include accessibility.min-tap-target();
    @include transition.add((visibility, opacity, transform));

    position: absolute; // 3.
    z-index: 3; // 3.
    display: flex;
    align-items: center;
    justify-content: center;
    color: var(--rui-local-arrow-color);
    visibility: hidden;
    opacity: 0;
}

.arrowIcon {
    @include caret.create();
}

.isRootVertical {
    height: 100%;
    min-height: 0; // 6.
}

.isRootVertical .viewport {
    height: 100%;
    overflow-x: hidden; // 7.
    overflow-y: auto; // 2.
}

.isRootVertical .arrowPrev {
    top: 0;
    right: 0;
    left: 0;
    width: 100%;
    padding-top: $_arrow-outer-spacing;
    padding-bottom: $_arrow-inner-spacing;
    transform: translateY(theme.$arrow-initial-offset);
}

.isRootVertical .arrowPrev .arrowIcon {
    @include caret.rotate(180);
}

.isRootVertical .arrowNext {
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    padding-top: $_arrow-inner-spacing;
    padding-bottom: $_arrow-outer-spacing;
    transform: translateY(calc(-1 * #{theme.$arrow-initial-offset}));
}

.isRootHorizontal {
    min-width: 0; // 6.
}

.isRootHorizontal .arrowPrev {
    top: 0;
    bottom: 0;
    left: 0;
    padding-right: $_arrow-inner-spacing;
    padding-left: $_arrow-outer-spacing;
    transform: translateX(theme.$arrow-initial-offset);
}

.isRootHorizontal .arrowPrev .arrowIcon {
    @include caret.rotate(90);
}

.isRootHorizontal .arrowNext {
    top: 0;
    right: 0;
    bottom: 0;
    padding-right: $_arrow-outer-spacing;
    padding-left: $_arrow-inner-spacing;
    transform: translateX(calc(-1 * #{theme.$arrow-initial-offset}));
}

.isRootHorizontal .arrowNext .arrowIcon {
    @include caret.rotate(270);
}

.isRootVertical .scrollingShadows::before,
.isRootVertical .scrollingShadows::after {
    right: 0;
    left: 0;
}

.isRootVertical .scrollingShadows::before {
    top: 0;
    transform: translateY(theme.$shadow-initial-offset);
}

.isRootVertical .scrollingShadows::after {
    bottom: 0;
    transform: translateY(calc(-1 * #{theme.$shadow-initial-offset}));
}

.isRootHorizontal .viewport {
    overflow-x: auto; // 2.
    overflow-y: hidden; // 5., 7.
}

.isRootHorizontal .content {
    display: inline-flex; // 4.
    min-width: 100%;
}

.isRootHorizontal .scrollingShadows::before,
.isRootHorizontal .scrollingShadows::after {
    top: 0;
    bottom: 0;
}

.isRootHorizontal .scrollingShadows::before {
    left: 0;
    transform: translateX(theme.$shadow-initial-offset);
}

.isRootHorizontal .scrollingShadows::after {
    right: 0;
    transform: translateX(calc(-1 * #{theme.$shadow-initial-offset}));
}

.isRootScrolledAtStart .scrollingShadows::before,
.isRootScrolledAtStart .arrowPrev {
    visibility: visible;
    opacity: 1;
    transform: translate(0, 0);
}

.isRootScrolledAtEnd .scrollingShadows::after,
.isRootScrolledAtEnd .arrowNext {
    visibility: visible;
    opacity: 1;
    transform: translate(0, 0);
}

.hasRootScrollbarDisabled .viewport {
    @include scrollbar.hide();
}
