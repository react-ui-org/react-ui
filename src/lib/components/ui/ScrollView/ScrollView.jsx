import PropTypes from 'prop-types';
import React, {
  useEffect,
  useRef,
  useState,
} from 'react';
import { useLoadResize } from '../../../hooks/useLoadResizeHook';
import { useScrollPosition } from '../../../hooks/useScrollPositionHook';
import { getElementsPositionDifference } from '../../../services/elementPositionService';
import { withTranslationContext } from '../../../translation';
import styles from './ScrollView.scss';

export const ScrollView = (props) => {
  const {
    arrows,
    arrowsColor,
    arrowsScrollStep,
    children,
    customEndShadowStyle,
    customNextArrow,
    customPrevArrow,
    customStartShadowStyle,
    id,
    debounce,
    direction,
    scrollbar,
    shadowColor,
    shadowSize,
    translations,
  } = props;

  const [isScrolledAtStart, setIsScrolledAtStart] = useState(false);
  const [isScrolledAtEnd, setIsScrolledAtEnd] = useState(false);

  const scrollPositionStart = direction === 'horizontal' ? 'left' : 'top';
  const scrollPositionEnd = direction === 'horizontal' ? 'right' : 'bottom';
  const scrollViewContentEl = useRef(null);
  const scrollViewViewportEl = useRef(null);

  const handleScrollViewState = (currentPosition) => {
    const isScrolledAtStartActive = currentPosition[scrollPositionStart] < 0;
    const isScrolledAtEndActive = currentPosition[scrollPositionEnd] > 0;

    if (isScrolledAtStartActive !== isScrolledAtStart) {
      setIsScrolledAtStart(isScrolledAtStartActive);
    }

    if (isScrolledAtEndActive !== isScrolledAtEnd) {
      setIsScrolledAtEnd(isScrolledAtEndActive);
    }
  };

  useEffect(
    () => {
      handleScrollViewState(
        getElementsPositionDifference(scrollViewContentEl, scrollViewViewportEl),
      );
    },
    [], // eslint-disable-line react-hooks/exhaustive-deps
  );

  useLoadResize(
    (currentPosition) => (handleScrollViewState(currentPosition)),
    [isScrolledAtStart, isScrolledAtEnd],
    scrollViewContentEl,
    scrollViewViewportEl,
    debounce,
  );

  useScrollPosition(
    (currentPosition) => (handleScrollViewState(currentPosition)),
    [isScrolledAtStart, isScrolledAtEnd],
    scrollViewContentEl,
    scrollViewViewportEl,
    debounce,
  );

  const setAlpha = (rgba, alpha) => ({
    ...rgba,
    alpha,
  });

  const rgbaToString = (rgba) => (
    `rgba(${rgba.red}, ${rgba.green}, ${rgba.blue}, ${rgba.alpha})`
  );

  const inlineStyle = (
    scrollDirection,
    customArrowsColor,
    customShadowColor,
    customShadowSize,
    startShadowStyle,
    endShadowStyle,
  ) => {
    const shadowStartColor = rgbaToString(customShadowColor);
    const shadowEndColor = rgbaToString(setAlpha(customShadowColor, 0));

    /* eslint-disable sort-keys */
    return {
      '--rui-local-arrow-color': customArrowsColor || 'inherit',
      '--rui-local-shadow-width': scrollDirection === 'horizontal' ? shadowSize : 'auto',
      '--rui-local-shadow-height': scrollDirection === 'horizontal' ? 'auto' : shadowSize,
      '--rui-local-start-shadow-background':
        startShadowStyle.background
        || `linear-gradient(${direction === 'horizontal' ? 'to right' : 'to bottom'}, ${shadowStartColor}, ${shadowEndColor})`,
      '--rui-local-start-shadow-box-shadow': startShadowStyle.boxShadow || 'none',
      '--rui-local-end-shadow-background':
        endShadowStyle.background
        || `linear-gradient(${direction === 'horizontal' ? 'to left' : 'to top'}, ${shadowStartColor}, ${shadowEndColor})`,
      '--rui-local-end-shadow-box-shadow': endShadowStyle.boxShadow || 'none',
    };
    /* eslint-enable sort-keys */
  };

  const arrowHandler = (contentEl, viewportEl, scrollViewDirection, shiftDirection, step) => {
    const offset = shiftDirection === 'next' ? step : -1 * step;
    const differenceX = scrollViewDirection === 'horizontal' ? offset : 0;
    const differenceY = scrollViewDirection !== 'horizontal' ? offset : 0;

    viewportEl.current.scrollBy(differenceX, differenceY);
  };

  return (
    <div
      className={[
        styles.root,
        isScrolledAtStart ? styles.isRootScrolledAtStart : '',
        isScrolledAtEnd ? styles.isRootScrolledAtEnd : '',
        scrollbar ? '' : styles.hasRootScrollbarDisabled,
        direction === 'horizontal' ? styles.isRootHorizontal : styles.isRootVertical,
      ].join(' ')}
      style={inlineStyle(
        direction,
        arrowsColor,
        shadowColor,
        shadowSize,
        customStartShadowStyle,
        customEndShadowStyle,
      )}
      {...(id && { id })}
    >
      <div className={styles.viewport} ref={scrollViewViewportEl}>
        <div
          className={styles.content}
          ref={scrollViewContentEl}
          {...(id && { id: `${id}__content` })}
        >
          {children}
        </div>
      </div>
      <div className={styles.scrollingShadows} aria-hidden />
      {arrows && (
        <>
          <button
            className={styles.arrowPrev}
            onClick={() => arrowHandler(
              scrollViewContentEl,
              scrollViewViewportEl,
              direction,
              'prev',
              arrowsScrollStep,
            )}
            title={translations.previous}
            {...(id && { id: `${id}__arrowPrevButton` })}
          >
            {customPrevArrow || (
              <span className={styles.arrowIcon} aria-hidden />
            )}
          </button>
          <button
            className={styles.arrowNext}
            onClick={() => arrowHandler(
              scrollViewContentEl,
              scrollViewViewportEl,
              direction,
              'next',
              arrowsScrollStep,
            )}
            title={translations.next}
            {...(id && { id: `${id}__arrowNextButton` })}
          >
            {customNextArrow || (
              <span className={styles.arrowIcon} aria-hidden />
            )}
          </button>
        </>
      )}
    </div>
  );
};

ScrollView.defaultProps = {
  arrows: false,
  arrowsColor: undefined,
  arrowsScrollStep: 200,
  customEndShadowStyle: {},
  customNextArrow: null,
  customPrevArrow: null,
  customStartShadowStyle: {},
  debounce: 50,
  direction: 'vertical',
  id: undefined,
  scrollbar: true,
  shadowColor: {
    alpha: 1,
    blue: 255,
    green: 255,
    red: 255,
  },
  shadowSize: '2em',
};

ScrollView.propTypes = {
  arrows: PropTypes.bool,
  arrowsColor: PropTypes.string,
  arrowsScrollStep: PropTypes.number,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  customEndShadowStyle: PropTypes.shape({
    background: PropTypes.string,
    boxShadow: PropTypes.string,
  }),
  customNextArrow: PropTypes.node,
  customPrevArrow: PropTypes.node,
  customStartShadowStyle: PropTypes.shape({
    background: PropTypes.string,
    boxShadow: PropTypes.string,
  }),
  debounce: PropTypes.number,
  direction: PropTypes.oneOf(['horizontal', 'vertical']),
  id: PropTypes.string,
  scrollbar: PropTypes.bool,
  shadowColor: PropTypes.shape({
    alpha: PropTypes.number.isRequired,
    blue: PropTypes.number.isRequired,
    green: PropTypes.number.isRequired,
    red: PropTypes.number.isRequired,
  }),
  shadowSize: PropTypes.string,
  translations: PropTypes.shape({
    next: PropTypes.string.isRequired,
    previous: PropTypes.string.isRequired,
  }).isRequired,
};

export default withTranslationContext(ScrollView, 'ScrollView');
