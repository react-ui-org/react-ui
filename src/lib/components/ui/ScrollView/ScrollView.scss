@import '../../../styles/tools/caret';
@import '../../../styles/tools/hide-scrollbar';
@import '../../../styles/tools/reset';
@import '../../../styles/tools/transitions';
@import 'theme';

// 1. Scrolling shadows are implemented as pseudo elements. This way we can customise them only
//    thanks to custom properties.
// 2. Clip scrolling shadows.
// 3. Create stacking context just to avoid scrolling shadows being covered by content.
// 4. Make the content bounding rectangle spread beyond the visible part.
// 5. Prevent undesired vertical scrolling that may occur with tables inside.

$_arrow-inner-offset: 0.5rem;
$_arrow-outer-offset: 1rem;

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  overflow: hidden; // 2.

  // 1.
  &::before,
  &::after {
    @include transition((visibility, opacity, transform));

    content: '';
    position: absolute;
    z-index: 1;
    display: block;
    visibility: hidden;
    opacity: 0;
    width: var(--rui-local-shadow-width);
    height: var(--rui-local-shadow-height);
  }

  &::before {
    background: var(--rui-local-start-shadow-background);
    box-shadow: var(--rui-local-start-shadow-box-shadow);
  }

  &::after {
    background: var(--rui-local-end-shadow-background);
    box-shadow: var(--rui-local-end-shadow-box-shadow);
  }
}

.viewport {
  position: relative; // 3.
  width: 100%;
  scroll-behavior: smooth;
  -webkit-overflow-scrolling: touch;
}

.arrowPrev,
.arrowNext {
  @include reset-button();
  @include transition((visibility, opacity, transform));

  position: absolute;
  z-index: 2;
  display: flex;
  visibility: hidden;
  opacity: 0;
  align-items: center;
  justify-content: center;
  color: var(--rui-local-arrow-color);
}

.arrowIcon {
  @include caret();
}

.isRootVertical {
  height: 100%;
}

.isRootVertical .viewport {
  height: 100%;
  overflow-y: auto;
}

.isRootVertical .arrowPrev {
  top: 0;
  right: 0;
  left: 0;
  width: 100%;
  padding-top: $_arrow-outer-offset;
  padding-bottom: $_arrow-inner-offset;
  transform: translateY($scrollview-arrow-initial-offset);
}

.isRootVertical .arrowPrev .arrowIcon {
  @include caret-rotate(180);
}

.isRootVertical .arrowNext {
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  padding-top: $_arrow-inner-offset;
  padding-bottom: $_arrow-outer-offset;
  transform: translateY(calc(-1 * #{$scrollview-arrow-initial-offset}));
}

.isRootHorizontal .arrowPrev {
  top: 0;
  bottom: 0;
  left: 0;
  padding-right: $_arrow-inner-offset;
  padding-left: $_arrow-outer-offset;
  transform: translateX($scrollview-arrow-initial-offset);
}

.isRootHorizontal .arrowPrev .arrowIcon {
  @include caret-rotate(90);
}

.isRootHorizontal .arrowNext {
  top: 0;
  right: 0;
  bottom: 0;
  padding-right: $_arrow-outer-offset;
  padding-left: $_arrow-inner-offset;
  transform: translateX(calc(-1 * #{$scrollview-arrow-initial-offset}));
}

.isRootHorizontal .arrowNext .arrowIcon {
  @include caret-rotate(270);
}

.isRootVertical::before,
.isRootVertical::after {
  right: 0;
  left: 0;
}

.isRootVertical::before {
  top: 0;
  transform: translateY($scrollview-shadow-initial-offset);
}

.isRootVertical::after {
  bottom: 0;
  transform: translateY(calc(-1 * #{$scrollview-shadow-initial-offset}));
}

.isRootHorizontal .viewport {
  overflow-x: auto;
  overflow-y: hidden; // 5.
}

.isRootHorizontal .content {
  display: inline-flex; // 4.
  min-width: 100%;
}

.isRootHorizontal::before,
.isRootHorizontal::after {
  top: 0;
  bottom: 0;
}

.isRootHorizontal::before {
  left: 0;
  transform: translateX($scrollview-shadow-initial-offset);
}

.isRootHorizontal::after {
  right: 0;
  transform: translateX(calc(-1 * #{$scrollview-shadow-initial-offset}));
}

.isRootScrolledAtStart::before,
.isRootScrolledAtStart .arrowPrev {
  visibility: visible;
  opacity: 1;
  transform: translate(0, 0);
}

.isRootScrolledAtEnd::after,
.isRootScrolledAtEnd .arrowNext {
  visibility: visible;
  opacity: 1;
  transform: translate(0, 0);
}

.hasRootScrollbarDisabled .viewport {
  @include hide-scrollbar();
}
