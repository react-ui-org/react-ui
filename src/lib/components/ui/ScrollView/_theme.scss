$arrow-initial-offset: var(--rui-ScrollView__arrow__initial-offset);
$shadow-initial-offset: var(--rui-ScrollView__shadow__initial-offset);
