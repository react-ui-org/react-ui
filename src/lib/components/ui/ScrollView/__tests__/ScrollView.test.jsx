import React from 'react';
import { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import * as elementPositionService from '../../../../services/elementPositionService';
import ScrollView from '..';

describe('rendering', () => {
  const mockedPositionDifference = {
    bottom: 0,
    left: 0,
    right: 0,
    top: 0,
  };
  const stub = sinon
    .stub(elementPositionService, 'getElementsPositionDifference')
    .returns(mockedPositionDifference);

  it('renders correctly with a single child', () => {
    const tree = mount((
      <ScrollView>
        <span>content</span>
      </ScrollView>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = mount((
      <ScrollView>
        <span>content 1</span>
        <span>content 2</span>
        <span>content 3</span>
      </ScrollView>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with scrollbar disabled', () => {
    const tree = mount((
      <ScrollView scrollbar={false}>
        <span>content</span>
      </ScrollView>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with arrows', () => {
    const tree = mount((
      <ScrollView arrows arrowsColor="white" arrowsScrollStep={200}>
        <span>content</span>
      </ScrollView>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = mount((
      <ScrollView
        arrows
        arrowsColor="blue"
        arrowsScrollStep={300}
        customEndShadow={{
          boxShadow: '0 0 10px rgba(0, 0, 0, 0.5)',
        }}
        customNextArrow={<span>→</span>}
        customPrevArrow={<span>←</span>}
        customStartShadow={{
          background: 'none',
        }}
        debounce={100}
        direction="horizontal"
        id="my-scrollview"
        scrollbar={false}
        shadowColor={{
          alpha: 0.5,
          blue: 122,
          green: 122,
          red: 122,
        }}
        shadowSize="100px"
      >
        <span>content 1</span>
        <span>content 2</span>
        <span>content 3</span>
      </ScrollView>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  stub.restore();
});
