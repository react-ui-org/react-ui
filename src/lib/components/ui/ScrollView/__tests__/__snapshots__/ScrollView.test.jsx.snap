// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<WithTranslationContextComponent
  translations={null}
>
  <ScrollView
    arrows={false}
    arrowsScrollStep={200}
    autoScroll="off"
    customEndShadowStyle={Object {}}
    customNextArrow={null}
    customPrevArrow={null}
    customStartShadowStyle={Object {}}
    debounce={50}
    direction="vertical"
    scrollbar={true}
    shadowColor={
      Object {
        "alpha": 1,
        "blue": 255,
        "green": 255,
        "red": 255,
      }
    }
    shadowSize="2em"
    translations={
      Object {
        "next": "Next",
        "previous": "Previous",
      }
    }
  >
    <div
      className="root    isRootVertical"
      style={
        Object {
          "--rui-local-arrow-color": "inherit",
          "--rui-local-end-shadow-background": "linear-gradient(to top, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-end-shadow-box-shadow": "none",
          "--rui-local-shadow-height": "2em",
          "--rui-local-shadow-width": "auto",
          "--rui-local-start-shadow-background": "linear-gradient(to bottom, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-start-shadow-box-shadow": "none",
        }
      }
    >
      <div
        className="viewport"
      >
        <div
          className="content"
        >
          <span>
            content
          </span>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="scrollingShadows"
      />
    </div>
  </ScrollView>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with all props 1`] = `
<WithTranslationContextComponent
  arrows={true}
  arrowsColor="blue"
  arrowsScrollStep={300}
  customEndShadow={
    Object {
      "boxShadow": "0 0 10px rgba(0, 0, 0, 0.5)",
    }
  }
  customNextArrow={
    <span>
      →
    </span>
  }
  customPrevArrow={
    <span>
      ←
    </span>
  }
  customStartShadow={
    Object {
      "background": "none",
    }
  }
  debounce={100}
  direction="horizontal"
  id="my-scrollview"
  scrollbar={false}
  shadowColor={
    Object {
      "alpha": 0.5,
      "blue": 122,
      "green": 122,
      "red": 122,
    }
  }
  shadowSize="100px"
  translations={null}
>
  <ScrollView
    arrows={true}
    arrowsColor="blue"
    arrowsScrollStep={300}
    autoScroll="off"
    customEndShadow={
      Object {
        "boxShadow": "0 0 10px rgba(0, 0, 0, 0.5)",
      }
    }
    customEndShadowStyle={Object {}}
    customNextArrow={
      <span>
        →
      </span>
    }
    customPrevArrow={
      <span>
        ←
      </span>
    }
    customStartShadow={
      Object {
        "background": "none",
      }
    }
    customStartShadowStyle={Object {}}
    debounce={100}
    direction="horizontal"
    id="my-scrollview"
    scrollbar={false}
    shadowColor={
      Object {
        "alpha": 0.5,
        "blue": 122,
        "green": 122,
        "red": 122,
      }
    }
    shadowSize="100px"
    translations={
      Object {
        "next": "Next",
        "previous": "Previous",
      }
    }
  >
    <div
      className="root   hasRootScrollbarDisabled isRootHorizontal"
      id="my-scrollview"
      style={
        Object {
          "--rui-local-arrow-color": "blue",
          "--rui-local-end-shadow-background": "linear-gradient(to left, rgba(122, 122, 122, 0.5), rgba(122, 122, 122, 0))",
          "--rui-local-end-shadow-box-shadow": "none",
          "--rui-local-shadow-height": "auto",
          "--rui-local-shadow-width": "100px",
          "--rui-local-start-shadow-background": "linear-gradient(to right, rgba(122, 122, 122, 0.5), rgba(122, 122, 122, 0))",
          "--rui-local-start-shadow-box-shadow": "none",
        }
      }
    >
      <div
        className="viewport"
      >
        <div
          className="content"
          id="my-scrollview__content"
        >
          <span>
            content 1
          </span>
          <span>
            content 2
          </span>
          <span>
            content 3
          </span>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="scrollingShadows"
      />
      <button
        className="arrowPrev"
        id="my-scrollview__arrowPrevButton"
        onClick={[Function]}
        title="Previous"
        type="button"
      >
        <span>
          ←
        </span>
      </button>
      <button
        className="arrowNext"
        id="my-scrollview__arrowNextButton"
        onClick={[Function]}
        title="Next"
        type="button"
      >
        <span>
          →
        </span>
      </button>
    </div>
  </ScrollView>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with arrows 1`] = `
<WithTranslationContextComponent
  arrows={true}
  arrowsColor="white"
  arrowsScrollStep={200}
  translations={null}
>
  <ScrollView
    arrows={true}
    arrowsColor="white"
    arrowsScrollStep={200}
    autoScroll="off"
    customEndShadowStyle={Object {}}
    customNextArrow={null}
    customPrevArrow={null}
    customStartShadowStyle={Object {}}
    debounce={50}
    direction="vertical"
    scrollbar={true}
    shadowColor={
      Object {
        "alpha": 1,
        "blue": 255,
        "green": 255,
        "red": 255,
      }
    }
    shadowSize="2em"
    translations={
      Object {
        "next": "Next",
        "previous": "Previous",
      }
    }
  >
    <div
      className="root    isRootVertical"
      style={
        Object {
          "--rui-local-arrow-color": "white",
          "--rui-local-end-shadow-background": "linear-gradient(to top, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-end-shadow-box-shadow": "none",
          "--rui-local-shadow-height": "2em",
          "--rui-local-shadow-width": "auto",
          "--rui-local-start-shadow-background": "linear-gradient(to bottom, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-start-shadow-box-shadow": "none",
        }
      }
    >
      <div
        className="viewport"
      >
        <div
          className="content"
        >
          <span>
            content
          </span>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="scrollingShadows"
      />
      <button
        className="arrowPrev"
        onClick={[Function]}
        title="Previous"
        type="button"
      >
        <span
          aria-hidden={true}
          className="arrowIcon"
        />
      </button>
      <button
        className="arrowNext"
        onClick={[Function]}
        title="Next"
        type="button"
      >
        <span
          aria-hidden={true}
          className="arrowIcon"
        />
      </button>
    </div>
  </ScrollView>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<WithTranslationContextComponent
  translations={null}
>
  <ScrollView
    arrows={false}
    arrowsScrollStep={200}
    autoScroll="off"
    customEndShadowStyle={Object {}}
    customNextArrow={null}
    customPrevArrow={null}
    customStartShadowStyle={Object {}}
    debounce={50}
    direction="vertical"
    scrollbar={true}
    shadowColor={
      Object {
        "alpha": 1,
        "blue": 255,
        "green": 255,
        "red": 255,
      }
    }
    shadowSize="2em"
    translations={
      Object {
        "next": "Next",
        "previous": "Previous",
      }
    }
  >
    <div
      className="root    isRootVertical"
      style={
        Object {
          "--rui-local-arrow-color": "inherit",
          "--rui-local-end-shadow-background": "linear-gradient(to top, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-end-shadow-box-shadow": "none",
          "--rui-local-shadow-height": "2em",
          "--rui-local-shadow-width": "auto",
          "--rui-local-start-shadow-background": "linear-gradient(to bottom, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-start-shadow-box-shadow": "none",
        }
      }
    >
      <div
        className="viewport"
      >
        <div
          className="content"
        >
          <span>
            content 1
          </span>
          <span>
            content 2
          </span>
          <span>
            content 3
          </span>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="scrollingShadows"
      />
    </div>
  </ScrollView>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with scrollbar disabled 1`] = `
<WithTranslationContextComponent
  scrollbar={false}
  translations={null}
>
  <ScrollView
    arrows={false}
    arrowsScrollStep={200}
    autoScroll="off"
    customEndShadowStyle={Object {}}
    customNextArrow={null}
    customPrevArrow={null}
    customStartShadowStyle={Object {}}
    debounce={50}
    direction="vertical"
    scrollbar={false}
    shadowColor={
      Object {
        "alpha": 1,
        "blue": 255,
        "green": 255,
        "red": 255,
      }
    }
    shadowSize="2em"
    translations={
      Object {
        "next": "Next",
        "previous": "Previous",
      }
    }
  >
    <div
      className="root   hasRootScrollbarDisabled isRootVertical"
      style={
        Object {
          "--rui-local-arrow-color": "inherit",
          "--rui-local-end-shadow-background": "linear-gradient(to top, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-end-shadow-box-shadow": "none",
          "--rui-local-shadow-height": "2em",
          "--rui-local-shadow-width": "auto",
          "--rui-local-start-shadow-background": "linear-gradient(to bottom, rgba(255, 255, 255, 1), rgba(255, 255, 255, 0))",
          "--rui-local-start-shadow-box-shadow": "none",
        }
      }
    >
      <div
        className="viewport"
      >
        <div
          className="content"
        >
          <span>
            content
          </span>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="scrollingShadows"
      />
    </div>
  </ScrollView>
</WithTranslationContextComponent>
`;
